File: book/HeadFirstDesignPatterns/go/ch5.singleton/chocolateboiler/ChocolateBoiler.go
package chocolateboiler

import (
	"fmt"
	"sync"
)

type chocolateBoiler struct {
	empty  bool
	boiled bool
	Val    int
}

var instance *chocolateBoiler
var once sync.Once

func GetInstance() *chocolateBoiler {
	once.Do(func() {
		instance = &chocolateBoiler{empty: true, boiled: false, Val: 0}
	})
	return instance
}

var mutex sync.Mutex

func GetInstance1() *chocolateBoiler {

	mutex.Lock()
	defer mutex.Unlock()
	if instance == nil {
		instance = &chocolateBoiler{empty: true, boiled: false, Val: 0}
	}
	return instance
}

func (c *chocolateBoiler) Fill() {
	if c.empty {
		c.empty = false
		c.boiled = false
		fmt.Println("초콜릿을 채운다")
	}
}

func (c *chocolateBoiler) Drain() {
	if c.empty == false && c.boiled {
		c.empty = true
		fmt.Println("초콜릿을 붓는다")
	}
}

func (c *chocolateBoiler) Boil() {

	if c.empty == false && c.boiled == false {
		c.boiled = true
		fmt.Println("초콜릿을 끓인다")
	}
}

File: book/HeadFirstDesignPatterns/java/ch5.singleton/src/ChocolateBoiler.java
public class ChocolateBoiler {
    private boolean empty;
    private boolean boiled;

    private volatile static ChocolateBoiler uniqueInstance = null;

    private ChocolateBoiler() {
            empty = true;
            boiled = false;
    }

    public static ChocolateBoiler getInstance() {
        if (uniqueInstance == null) {
            synchronized (ChocolateBoiler.class) {
                if (uniqueInstance == null) {
                    uniqueInstance = new ChocolateBoiler();
                }
            }
        }
        return uniqueInstance;
    }

   public void fill() {
         if (isEmpty()) {
            empty = false;
            boiled = false;
            System.out.println("초콜릿을 채운다");
        }
    }

    public void drain() {
        if (!isEmpty() && isBoiled()) {
            //끓인 재료를 다음 단계로 넘김
            empty = true;
            System.out.println("초콜릿을 붓는다");
        }
    }

    public void boil() {
        if (!isEmpty() && !isBoiled()) {
            //재료를 끓임
            boiled = true;
            System.out.println("초콜릿을 끓인다");
        }
    }

    public boolean isEmpty() {
        return empty;
    }

    public boolean isBoiled() {
        return boiled;
    }
}
